src/lib/support/StateMachine.h

```diff
@@ -48,37 +48,40 @@ struct VariantState : Variant<Ts...>
 
 private:
     template <typename T>
-    void Enter()
+    void Enter(bool & ever)
```

Comment:
What is the purpose of the `ever` argument?

Reply:
This is introduced by this commit: https://github.com/project-chip/connectedhomeip/pull/13294/commits/4f3b19b1945b9e80e31995e2722e74fcd8858571

It turns out that because Enter() can destruct VariantState _this_, but also constructs a potentially different state in place, there comes about a situation where the pattern match can evaluate true multiple times.  This boolean prevents that.  This is preventing further access to object member variables after the object may have been destructed.

The behavior where Enter() destructs the state while other states may be recursively constructed on the stack was introduced to remove the deque, per the discussion here: https://github.com/project-chip/connectedhomeip/pull/12223#discussion_r757060458.

Reply:
This is now addressed in a separate PR here: https://github.com/project-chip/connectedhomeip/pull/13664